src/expand.c

```diff
@@ -0,0 +1,138 @@
+#include "vctrs.h"
+#include "type-data-frame.h"
+
+#include "decl/expand-decl.h"
+
+r_obj* ffi_vec_expand_grid(r_obj* xs,
+                           r_obj* ffi_vary,
+                           r_obj* ffi_name_repair,
+                           r_obj* frame) {
+  struct r_lazy error_call = { .x = syms.dot_error_call, .env = frame };
+
+  enum vctrs_expand_vary vary = parse_vary(ffi_vary);
+
+  struct name_repair_opts name_repair_opts = new_name_repair_opts(
+    ffi_name_repair,
+    lazy_args.dot_name_repair,
+    false,
+    error_call
+  );
+  KEEP(name_repair_opts.shelter);
+
+  r_obj* out = vec_expand_grid(xs, vary, &name_repair_opts, error_call);
+
+  FREE(1);
+  return out;
+}
+
+r_obj* vec_expand_grid(r_obj* xs,
+                       enum vctrs_expand_vary vary,
+                       const struct name_repair_opts* p_name_repair_opts,
+                       struct r_lazy error_call) {
+  vec_check_list(xs, vec_args.empty, error_call);
+
+  if (vec_any_missing(xs)) {
+    // Drop `NULL`s before any other checks
+    r_obj* complete = KEEP(vec_detect_complete(xs));
+    xs = vec_slice(xs, complete);
+    FREE(1);
+  }
```

Comment:
We thought about exposing `list_drop_missing()` (or `vec_drop_missing()`, I guess), which would be a complement to `list_drop_empty()` (that one really does need a list_ prefix because only lists hold empty vectors) https://github.com/r-lib/vctrs/issues/1563

In the end I decided that using `vec_any_missing() + vec_detect_missing()` was a more flexible pattern, and I almost wish I had exposed `list_any_empty()` and `list_detect_empty()` instead of `list_drop_empty()`, because tidyr could have used those individual pieces